packages/asset-swapper/contracts/src/SmoothySampler.sol

```diff
@@ -0,0 +1,179 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6;
+pragma experimental ABIEncoderV2;
+
+// import "./interfaces/ISmoothy.sol";
+import "./ApproximateBuys.sol";
+import "./SamplerUtils.sol";
+import "./interfaces/ISmoothy.sol";
+
+contract SmoothySampler is
+    SamplerUtils,
+    ApproximateBuys
+{
+    /// @dev Information for sampling from smoothy sources.
+    struct SmoothyInfo {
+        address poolAddress;
+        bytes4 sellQuoteFunctionSelector;
+        bytes4 buyQuoteFunctionSelector;
+    }
+
+    /// @dev Base gas limit for Smoothy calls.
+    uint256 constant private SMOOTHY_CALL_GAS = 600e3;
+
+    /// @dev Sample sell quotes from Smoothy.
+    /// @param smoothyInfo Smoothy information specific to this token pair.
+    /// @param fromTokenIdx Index of the taker token (what to sell).
+    /// @param toTokenIdx Index of the maker token (what to buy).
+    /// @param takerTokenAmounts Taker token sell amount for each sample.
+    /// @return makerTokenAmounts Maker amounts bought at each taker token
+    ///         amount.
+    function sampleSellsFromSmoothy(
+        SmoothyInfo memory smoothyInfo,
+        int128 fromTokenIdx,
+        int128 toTokenIdx,
+        uint256[] memory takerTokenAmounts
+    )
+        public
+        view
+        returns (uint256[] memory makerTokenAmounts)
+    {
+        // Basically a Curve fork 
+
+        // Smoothy only keep a percentage of its tokens available in reserve
+        uint256 poolReserveMakerAmount = ISmoothy(smoothyInfo.poolAddress).getBalance(uint256(toTokenIdx)) - 
+                                         ISmoothy(smoothyInfo.poolAddress)._yBalances(uint256(toTokenIdx));
+        (uint256 softWeight, uint256 hardWeight, uint256 balance, uint256 decimals) = 
+                                         ISmoothy(smoothyInfo.poolAddress).getTokenStats(uint256(toTokenIdx));
+
+        uint256 numSamples = takerTokenAmounts.length;
+        makerTokenAmounts = new uint256[](numSamples);
+        for (uint256 i = 0; i < numSamples; i++) {
+            (bool didSucceed, bytes memory resultData) =
+                smoothyInfo.poolAddress.staticcall.gas(SMOOTHY_CALL_GAS)(
+                    abi.encodeWithSelector(
+                        smoothyInfo.sellQuoteFunctionSelector,
+                        fromTokenIdx,
+                        toTokenIdx,
+                        takerTokenAmounts[i]
+                    ));
+            uint256 buyAmount = 0;
+            if (didSucceed) {
+                buyAmount = abi.decode(resultData, (uint256));
+            }
+
+            // Make sure the quoted buyAmount is available in the pool reserve 
+            if (buyAmount*(10**(18-decimals)) >= poolReserveMakerAmount) {
```

Comment:
Smoothy keeps some assets as y tokens (for example yUSDT). Sometimes there are no more USDT in the pool but only yUSDT. Problem was the yUSDT vault had some issue where the USDT was "locked" because it was borrowed and couldn't be withdrawn. The `getSwapAmount` quotes us assuming the USDT is withdrawable, so in the sampler I made sure we only look at that's in the contract (without checking yUSDT) so it can never revert. This is to be on the safe side. Also withdrawing from yUSDT would add a lot of gas overhead so not worth it.

Reply:
https://smoothy-finance.medium.com/smoothy-finance-withdrawal-incident-analysis-d9a85ab52983

Reply:
https://etherscan.io/address/0xe5859f4efc09027a9b718781dcb2c6910cac6e91